### PR TITLE
feat:profiling:state summary and visualization

### PR DESCRIPTION
**Pull request by ZenGround0**
## Related Issues

#10884 

## Proposed Changes

* One lotus-shed command that takes a non-running repo and generates a summary of data usage by protocol area
* A python script for visualizing the data 

### Testplan

I've done many manual runs on calibration and mainnet repos and both lotus-shed and python script generate plausible output.  Ive cross checked output data and it appears legit matching lotus chain store size + providing an explanation for pre nv19 snapshot bloat.  I think this is good enough for running a monitoring service.  If we find problems we can fix them as they show up.

### Speed

While reviewing please keep an eye out for optimization opportunities in the shed command, today it takes about 5 hours to run so a 20% improvement is an hour saved.

### Halp

@jennijuju where should documentation for this go?  I have lotus-shed command help but probably we want something a little more thorough.  Is https://github.com/filecoin-project/lotus/discussions/categories/tutorials enough?

### Extension ideas

I am hoping to tweak the python script to take in visualization chart type information so we can visualize same data in a few different ways.

State summary shed command is readily extensible by using a path identifier for protocol areas.  We could extend this to breakdown message usage by message type for example.

We could use similar visualization on gas traces.  In particular cron gas traces would be useful to keep track of in the same way.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message. (they will after squash)
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: cmd/lotus-shed/state-stats.go [hunk 1]
```diff
@@ -166,49 +186,393 @@ var statObjCmd = &cli.Command{
...
+					return
 				}
 			}
 		}()
+		visit := func(c cid.Cid) bool {
+			ch := make(chan bool)
+			cidCh <- cidCall{c: c, resp: ch}
+			out := <-ch
+			return out
+		}
```
Comment (magik6k) [inline cmd/lotus-shed/state-stats.go:400]: I'd just use a mutex, should me much faster

File: cmd/lotus-shed/state-stats.go [hunk 2]
```diff
@@ -166,49 +186,393 @@ var statObjCmd = &cli.Command{
...
+}
+
+type StoreHandle struct {
+	bs     blockstore.Blockstore
+	cs     *store.ChainStore
+	sm     *stmgr.StateManager
+	closer func()
+}
+
+func loadChainStore(ctx context.Context, repoPath string) (*StoreHandle, error) {
```
Comment (magik6k) [inline cmd/lotus-shed/state-stats.go:216]: Probably could have a mode where this loads a snapshot to a memory-map blockstore, would need a decent amount of RAM, but should be much faster

**Comment by snissn** [comment]
Having this in two branches is causing an issue on the deployment infra. I think this code is good to merge, and all of the comments above are good suggestions that we should open a second issue for.

**Comment by ZenGround0** [comment]
> I think this code is good to merge

yup, I just need a green check mark, can you approve @snissn ?

File: cmd/lotus-shed/state-stats.go (hunk 1 above)
Comment (ZenGround0) [reply cmd/lotus-shed/state-stats.go:400]: I tried it out, was minorly faster but this is apparently not a bottleneck

File: cmd/lotus-shed/state-stats.go [hunk 3]
```diff
@@ -427,6 +748,93 @@ to reduce the number of decode operations performed by caching the decoded objec
...
+		results = append(results, res)
+	}
+
+	// now walk the top level object of actor state
+	dsc := &dagStatCollector{
+		ds:   dag,
+		walk: carWalkFunc,
+	}
+
+	if err := merkledag.Walk(ctx, dsc.walkLinks, in.c, visit, merkledag.Concurrent()); err != nil {
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/lotus-shed/state-stats.go:831]: this PR has been working well in testing, @ZenGround0 if you can investigate this, but i don't think we're bottle necked here

**Review by snissn — APPROVED** [review]
looks good to me, has been working and generated expected output for us!

**Review by arajasek — COMMENTED** [review]
Sorry, I never submitted this (partial) review!

**Pull request merged**